common_benchmark_suite/openxla/benchmark/definition_types.py

```diff
@@ -0,0 +1,136 @@
+  BF16 = "bf16"
+  INT8 = "int8"
+  UINT8 = "uint8"
+
+
+@dataclass(frozen=True)
+class ModelImplementation:
+  """Model implementation with configurable parameters."""
+  id: str
+  # Friendly unique name.
+  name: str
+  # Tags that describe the model characteristics.
+  tags: List[str]
+  # Framework the model is implemented in.
+  framework_type: ModelFrameworkType
+  # Data type used in the model.
+  data_type: ModelDataType
+  # Source of the model.
+  source_info: str
+
+  def __str__(self):
+    return self.name
+
+
+class ModelArtifactType(Enum):
+  """Type of derived model artifact."""
+  TF_SAVEDMODEL_V1 = "tf_savedmodel_v1"
+  TF_SAVEDMODEL_V2 = "tf_savedmodel_v2"
+  XLA_HLO_DUMP = "xla_hlo_dump"
+  STABLEHLO = "stablehlo"
+
+
+@dataclass(frozen=True)
+class ModelInstance:
+  """A model instance with concrete parameters to initialize the model."""
+  id: str
+  # Friendly unique name.
+  name: str
+  # Tags that describe the characteristics additional to model_impl.tags.
+  tags: List[str]
+  # Source model implementation.
+  model_impl: ModelImplementation
+  # Parameters to initialize the model, e.g., input batch size, sequence length.
+  model_parameters: Dict[str, Any]
+  # URLs to download the derived models of the initialized model.
+  artifact_sources: Dict[ModelArtifactType, str]
+
+  def __str__(self):
+    return self.name
+
+
+class ModelTestDataFormat(Enum):
+  """Model input or output data format."""
+  # Pack of npy tensor files.
+  NUMPY_TENSORS = "npy_tensors"
+
+
+@dataclass(frozen=True)
+class ModelTestData:
+  """Model input or expected output data."""
+  id: str
+  # Friendly name.
+  name: str
+  # Tags that describe the data characteristics.
+  tags: List[str]
+  # Information on where the data was originally sourced.
+  source_info: str
+  # URLs to download the data in multiple formats.
+  data_sources: Dict[ModelTestDataFormat, str]
+  # Parameters for output verifiers if applicable.
+  output_verify_params: Dict[ModelTestDataFormat, Dict[str, Any]]
+
+  def __str__(self):
+    return self.name
+
+
+@dataclass(frozen=True)
+class DeviceSpec:
+  """Device specification to run benchmarks."""
+  id: str
+  # Friendly unique name.
+  name: str
+  # Describes the host that runs the runtime and talks to the accelerator.
+  # E.g., GCP
+  host_type: str
+  # E.g., c2-standard-60
+  host_model: str
+  # E.g., linux-x86_64
+  host_environment: str
+
+  # Describes the target accelerator (can be the same as the host for CPU
+  # benchmarks).
+  # E.g., cpu, gpu, my-custom-accelerator
+  accelerator_type: str
+  # E.g., nvidia-a100-40g
+  accelerator_model: str
+  # E.g., intel-cascadelake, nvidia-ampere
+  accelerator_architecture: str
+  # E.g., "num_of_gpus": 4, "cpu_mask": "0-3"
+  accelerator_attributes: Dict[str, Any]
+
+  def __str__(self):
+    return self.name
+
+
+@dataclass(frozen=True)
+class BenchmarkInstance:
+  """A benchmark instance."""
+  model: ModelInstance
+  input_data: ModelTestData
```

Comment:
The model may have multiple inputs and outputs, so `input_data` and `expected_output` should be a list or `ModelTestData` should contain the list.

Reply:
I'm thinking maybe we can pack multiple input/output npy into a `.tgz` to not expose the test data format in the common `BenchmarkInstance`. It is demonstrated here https://github.com/openxla/openxla-benchmark/pull/9/files#diff-8b4b6206e9c4d722b52446003f2044e6d62bbb3ae99e7d5aa303e830b2b6ad2d